### PR TITLE
Use DataFrames for ChainSummaries

### PR DESCRIPTION
**Pull request by cpfiffer**
This PR switches MCMCChains off the strange I/O driven `ChainSummary` and `ChainSummaries` structs, and uses a much simpler `ChainDataFrame` struct that brings a lot of the versatility of `DataFrames` for printing, organization, and indexing. It's now a lot easier to get values out of `ChainSummary` structs (see #71).

Currently, the bulk of the work is done. There's still a couple of remaining steps, namely tidying and making sure all the tests are doing what they're supposed to:

- [x] Remove `ChainSummary` and `ChainSummaries` entirely
- [x] Make all the tests pass

There is a new function called `summarize` which maps a function parameter-wise and returns a dataframe. Hopefully this'll make it quicker to get some non-traditional summary stats, and it's used a lot on the backend. Example:

```julia
using MCMCChains, Statistics

chn = Chains(rand(500, 3, 2))
x = summarize(chn, mean, var, std)
```
```
│ Row │ parameters │ mean     │ var       │ std      │
│     │ Symbol     │ Float64  │ Float64   │ Float64  │
├─────┼────────────┼──────────┼───────────┼──────────┤
│ 1   │ Param1     │ 0.49674  │ 0.0877798 │ 0.296277 │
│ 2   │ Param2     │ 0.492215 │ 0.0807893 │ 0.284235 │
│ 3   │ Param3     │ 0.49734  │ 0.0850419 │ 0.291619 │
```

The indexing on these is overloaded to put a higher priority on rows:

```julia
x[:mean] # Select the mean column
x[:Param1, :mean] # select the mean for Param1
x[:Param1, :] # Select the Param1 row
```

The `show` function for `Chains` now uses this as well:

```julia
julia> chn
Object of type Chains, with data of type 500×3×2 Array{Float64,3}

Iterations        = 1:500
Thinning interval = 1
Chains            = 1, 2
Samples per chain = 500
parameters        = Param1, Param2, Param3

Summary Statistics

│ Row │ parameters │ mean     │ std      │ naive_se   │ mcse       │ ess     │
│     │ Symbol     │ Float64  │ Float64  │ Float64    │ Float64    │ Float64 │
├─────┼────────────┼──────────┼──────────┼────────────┼────────────┼─────────┤
│ 1   │ Param1     │ 0.49674  │ 0.296277 │ 0.00936909 │ 0.0109575  │ 731.097 │
│ 2   │ Param2     │ 0.492215 │ 0.284235 │ 0.00898829 │ 0.00939995 │ 914.329 │
│ 3   │ Param3     │ 0.49734  │ 0.291619 │ 0.00922182 │ 0.00675694 │ 1000.0  │
```
All the stats functions include keyword arguments for `sections` and `append_chains`. If `append_chains=true`, then all the chains are smushed together into one chain. Otherwise, you'll get a vector of `ChainDataFrames`, one for each chain. `sections` just allows you to subset your chain by sections.

### CODE REVIEW

**Comment by cpfiffer** [comment]
It looks like we've been calculating ESS using 

```julia
min((std(x) / df_mcse(x))^2, length(x))
```

which seems really strange to me. Shouldn't we use something more like what [Stan](https://mc-stan.org/docs/2_18/reference-manual/effective-sample-size-section.html) (and lots more) use?

If nobody is opposed, I'm going to update our ESS calculator to use the autocorrelation-derived variant.

**Comment by goedman** [comment]
Hi Cameron, I've been looking at https://arxiv.org/abs/1903.08008 . Have you seen that update?

**Comment by cpfiffer** [comment]
I hadn't, no. Funnily enough I was sitting down with the "Bayesian Data Analysis" book by Gelman et al. (2014) which implements this ESS.

Maybe we should open a separate issue to address some of the points brought up in the article you mentioned.

**Comment by goedman** [comment]
Good idea. 

I first want to look at WAIC and that modelchain stuff in MCMCChains. 

Also of interest on LOO etc might be, a bit older ( http://www.jmlr.org/papers/volume11/watanabe10a/watanabe10a.pdf ). Maybe we should make that a separate issue as well.

**Comment by cpfiffer** [comment]
I've added an `ess` function which liberally lifts from the [Stan implementation](https://github.com/stan-dev/stan/blob/11d5b7b8d3edbc2dfd4883859e741ec9b156aa5e/src/stan/analyze/mcmc/compute_effective_sample_size.hpp#L25). It'll show up now whenever `describe`, `summarystats`, or `show(chn)` is called.

**Comment by cpfiffer** [comment]
Looks like our tests are passing. This is now ready for a formal review.

**Comment by goedman** [comment]
I think this looks great Cameron!

**Comment by trappmartin** [comment]
I'm happy to review the PR by tomorrow.

**Comment by goedman** [comment]
Cameron, just as a thought, would it be useful to have `describe()` return the summary data frame?
Working through the CmdStan examples I think I'll end up calling both most of the time.

**Comment by cpfiffer** [comment]
I actually wasn't sure what to do about that, so I'm happy to take suggestions there. My thinking as to why it doesn't actually return anything is that I had thought it was intended as primarily an I/O function to show you information about the chain --- not to return the dataframes. I had it set to do both but if you ran `describe` from the REPL it would print the value out twice: one for the `show` calls in the function, and one for the return value.

Perhaps a better option would be to have `describe(chn)` only return the dataframes (with no printing or internals `show` calls), and then set `show(chn)` to call `describe` in the back end. Does that sounds slightly more reasonable?

File: src/discretediag.jl [hunk 1]
```diff
@@ -416,11 +416,60 @@ end
 #  return [p1, p2]
 #end
 
+# function discretediag(chn::AbstractChains; frac::Real=0.3,
```
Comment (trappmartin) [inline src/discretediag.jl]: Cant this be removed?

File: src/discretediag.jl [hunk 2]
```diff
@@ -440,15 +489,18 @@ function discretediag(chn::AbstractChains; frac::Real=0.3,
     V, vals = discretediag_sub(c, frac, method, nsim,
     size(c.value,1), size(c.value,1))[1:2]
 
-    #println(V)
-    #println(vals)
+    num_chains = length(chains(c))
+    sep_vals = [vals[(3 * (k - 1) + 1):(3 * (k - 1) + 3), :]'
```
Comment (trappmartin) [inline src/discretediag.jl]: Maybe add a comment what is happening here.

**Comment by trappmartin** [comment]
Look good to me. I would not add any further functionalities as PR is already quite large. Better merge the PR and open a new one for the `describe` function.

**Comment by goedman** [comment]
Yes, running it in the REPL would require `s = describe(chns);` I guess, with the `;`. I could live with that. Part of my suggestion is to prevent, for very big models, generating the data frame twice. But in general the time the sampling takes dominates the post processing anyway.

**Comment by cpfiffer** [comment]
Merging for now, since I've only modified comments and previous tests were passing. Thanks for the review, guys!

**Pull request merged**